SlotFill: some code cleanups of the bubblesVirtually version

jsnajdr opened the pull request:
After the successful #50098, here are some more code improvements in the `bubblesVirtually` version of `SlotFill`:
- when creating the `registry` for the context, we can save many `useRef`s and `useCallback`s by creating the registry as a constant object and then storing it, forever, in provider's state.
- similarly, we can save many `useCallback`s in `useSlot` by creating the bound API object as a single `api` object, with `useMemo`.
- finally there's a tiny change in `Fill` to move hook call into a separate statement, as is the common custom.

jsnajdr committed:
SlotFillProvider: simplify creating registry

jsnajdr committed:
useSlot: simplify creating the bound API object

jsnajdr committed:
Fill: call hook in its own statement

Mamaduka reviewed (commented):
The SlotFills "package" also has two hooks named `useSlot` one is a [public API](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/slot-fill/bubbles-virtually/use-slot.js) and another [used internally](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/slot-fill/use-slot.js). It would be nice to rename latter and avoid confusion.

tyxla reviewed (approved):
The cleanup looks good IMHO 👍 Thanks!

I agree with @Mamaduka that the naming conflict can be confusing, but I'd say that the entire situation with the two "standard" and "bubbles virtually" versions is a bit awkward too. Ideally, we'd have a single version that we could just configure. 

Optionally, this could have a CHANGELOG entry, but up to you really.

jsnajdr commented:
> two hooks named `useSlot` one is a public API

There are just two usages of `useSlot`, one of them I'm removing in #50198. The other is in `Popover`, and I suspect it's removable, too. Then we could deprecate this public API, and make `useSlot` an internal helper again.

I'm also wondering if we can somehow get rid of `useSlotFills`, too. The only use case for it is to check whether a slot has any registered fills, looking just at the yes/no boolean information, never using the actual fills data. This information is then used to conditionally render the `Slot`.

jsnajdr committed:
Added changelog entry

jsnajdr commented:
> Optionally, this could have a CHANGELOG entry

Added the changelog entry in 7818c261ce, both for this PR and the related #50098.

tyxla commented:
> I'm also wondering if we can somehow get rid of `useSlotFills`, too. The only use case for it is to check whether a slot has any registered fills, looking just at the yes/no boolean information, never using the actual fills data. This information is then used to conditionally render the `Slot`.

Well, `registry.fills` is a `Map` ([`ProxyMap`](https://valtio.pmnd.rs/docs/api/utils/proxyMap) to be precise) and while we could theoretically use `registry.fills.size`, I think there's a value in using the hook, since it uses `useSnapshot` to optimize for updates only when necessary - as per [docs](https://valtio.pmnd.rs/docs/api/basic/useSnapshot):

> `useSnapshot` wraps the Valtio snapshot in an access-tracking proxy, so your component is render optimized, i.e. it will only re-render if keys that it (or its child components) specifically accessed have changed, and not on every single change to the proxy.

That being said, I'd rather keep the hook and benefit from the pre-existing performance optimizations. We could of course do some measurements.

jsnajdr commented:
> there's a value in using the hook, since it uses `useSnapshot` to optimize for updates only when necessary

Yes, I'd like to keep all performance benefits, I'm just thinking aloud about what we really need `useSlotFills` for, and if it has the ideal API for that. It seems that the sole use case is to render a `Slot` or some other UI (inspector tab) only if we have any actual fills to fill the UI with. For that purpose, returning a `boolean` would be enough. Also, this use case is valid also for the v1 slots, with `bubblesVirtually=false`, so `useSlotFills` should work for both versions.

About `valtio` usage: I think that with modern `useSyncExternalStore`, we could implement the "trigger rerenders only when actually accessed keys change" optimization also in vanilla JS/React. It's much easier now than it was before.

jsnajdr commented:
The "Verify changelog update" CI check is still unhappy about the changelog. Tt's because I added an entry for two PRs at once, referencing both of them:
```
- `SlotFill`: simplified the implementation and removed unused code (#50098 and #50133)
```
But the regexp in the check is more strict about parentheses, accepting only the `(#50098)` form. Cc @chad1008 -- maybe we want to relax the rules a bit? Not sure how frequently such changelog entries are done.

chad1008 commented:
Good point @jsnajdr, thanks for the ping. I see why it's confused and it'll be an easy enough fix. I'll get a PR together.
